calc/src/mechanics/gen2.ts

```diff
@@ -127,6 +127,18 @@ export function calculateGSC(attacker: Pokemon, defender: Pokemon, move: Move, f
     at = Math.floor(at / 4) % 256;
     df = Math.floor(df / 4) % 256;
   }
+  
+  // Gen 2 Present has a glitched damage calculation using the secondary types of the Pokemon for the Attacker's Level and Defender's Defense.
+  if (move.name === 'Present') {
```

Comment:
This doesn’t seem like the right calculation? Also can you write tests?

Reply:
The calculation is correct. When Present is used, after the Attack/Defense/Level values are read from and stored, they are overwritten by Present's buggy code.

The "at"/Attack value is set to 10.
The "df"/Defense value becomes the Attacker's secondary type index.
The "lv"/Level value becomes the Defender's secondary type index.
Mono-type Pokemon have the same type twice internally, so these Pokemon have a primary and secondary type of 'Normal', 'Water', etc.

This means that when used by a Normal-type, the attack has 10 Attack against a 0 Defense target (Normal's index = 0, Attacker's type = Normal, Defender's Defense is set to Attacker's type which is 0), with the type index of the Defender used in place of the Attacker's level, which is lower for 'physical'-types (Normal=0, Fighting=1, etc.) and higher for 'special'-types (Fire=20, Water=21, etc.). This makes the attack unreasonably powerful against Pokemon such as Mewtwo, Mew, Celebi, or any other Pokemon with a secondary type index of 20 or higher (any 'special' type), to the point a Level 5 Togepi has the capacity to OHKO a level 100 Umbreon as shown in Crystal_'s Youtube video if it gets a 120 BP Present (1/10 Chance).

This is what my modification to the calculator will show for Present:
Togepi (set to Level 5) Present (set to 120 BP) vs. Umbreon (set to Level 100): 370-435 (94.1 - 110.6%) -- 64.1% chance to OHKO after Leftovers recovery

This happens in Crystal_'s video at the point Togepi nearly OHKOs the Umbreon. Due to Present overwriting the Attacker's Attack/Attacker's Level/Defender's Defense used for calculation, Togepi's attack becomes a level 27 (Dark's index is 27) STAB 40/80/120 BP attack (or a 1/4 Heal) with an Attack stat of 10 against a Defense stat of 0. This overwrite happens *after* boosts, screens and so on are applied to Attacker's Attack/Defender's Defense, so Present ignores the Attack stage of the attacker, as well as the Defense stages and screens of the defender.

Using Showdown's GSC damage calculation shows that:
let damage = level[27] * 2 -- damage = 54
damage = Math.floor(damage[54] / 5) -- damage = 10
damage += 2 -- damage = 12
damage *= 120 (max damage of Present) -- damage = 1440
damage *= attack[10] -- damage = 14400
damage = Math.floor(damage[14400] / defense[0]) --
damage = clampIntRange(Math.floor(damage[14400] / 50) = 288, 1, 997) -- damage = 288
damage += 2 -- damage = 290

--STAB Bonus--
damage += Math.floor(damage / 2) -- damage = 435

435 = max damage of STAB 120BP Present from a mono Normal-type vs. Umbreon, same as my calc shows.
435 * 217 / 255 = 370, the minimum damage of STAB 120BP Present from a mono Normal-type vs. Umbreon, same as my calc shows.

Reply:
Hm alright, the coding did the actual calculation looked weird fsr. Can you write tests though?

Reply:
"Write tests"? Not sure what you mean. I thought you meant showing that the results were correct, which I did (the Togepi and Togetic calcs came from tests from index.html after applying my changes), but you seem to want something else?

Reply:
No we actually have test, see https://github.com/smogon/damage-calc/blob/master/calc/src/test/move.test.ts

Reply:
If I understand correctly, I needed to do a npm test? It gave me some issues when it got to the gts check at first, but I cleared them up and passes through the test fine. Or am I supposed to do something else?

Reply:
No not that but yeah you should always `/npm test` before making a pr

Reply:
See the commit I linked for the type of tests I’m talking about